File: python_july/124. Binary Tree Maximum Path Sum.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        self.max_sum = -float('inf')
        self.helper(root)
        return self.max_sum

    def helper(self, node):
        if not node:
            return 0
        
        leftSum = max(self.helper(node.left), 0)
        rightSum = max(self.helper(node.right), 0)

        inorder = node.val + leftSum + rightSum
        self.max_sum = max(self.max_sum, inorder)

        return node.val + max(leftSum, rightSum)

File: java_august/124. Binary Tree Maximum Path Sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    private int maxSum;

    public int maxPathSum(TreeNode root) {
        maxSum = Integer.MIN_VALUE;
        getMax(root);
        return maxSum;
    }

    private int getMax(TreeNode node) {
        if (node == null) {
            return 0;
        }
        int left = Math.max(getMax(node.left), 0);
        int right = Math.max(getMax(node.right), 0);

        maxSum = Math.max(maxSum, node.val + left + right);

        return node.val + Math.max(left, right);
    }
}